Add CMS AuthEnvelopedData with AES-GCM support

bukka opened the pull request:
This PR adds support for CMS AuthEnvelopedData as defined in RFC 5083 with AES-GCM parameter as defined in RFC 5084.

I think it useful to support GCM as it allows the tag to be stored with the message. Especially if the cms is now recommended tool for AEAD and there was a decision not to introduce special format for apps enc to handle AEAD modes IIRC.

It works already with use of apps cms when selecting for example `-aes-128-gcm` cipher. It updates `CMS_encrypt` to automatically use AuthEnvelopedData for GCM modes so the application doesn't require new API. As there is now new special API, it means that there is currently no support for AAD but that's something that I would like to look in the future as part of another PR. Also CCM mode is not supported as it's a bit more tricky (mainly because the length needs to be known in advance) but I plan to look into it in the future too.

The documentation is partially updated which means that I updated `CMS_encrypt` and `CMS_decrypt`. The reason why I left the rest out is that I'm not sure if the new public functions should be really public. Currently I added them as a public for consistency. For example `EVP_CIPHER_set_asn1_aead_params` is an AEAD variant for `EVP_CIPHER_set_asn1_iv` so it might be useful to expose it possibly for engines use. Similarly `ASN1_TYPE_set_octetstring_int` is consistent with public `ASN1_TYPE_set_int_octetstring` (also not documented). In addition the new `CMS_AuthEnvelopedData_create` is like exposed `CMS_EnvelopedData_create` (also not documented) but for AuthEnvelopedData of course. Considering that all the existing mentioned functions are not documented, it might be better to leave the documentation for another PR and do it together with the existing once. I would be happy to do that later.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

FdaSilvaYY commented on crypto/cms/cms_env.c:
```diff
@@ -48,6 +79,25 @@ static CMS_EnvelopedData *cms_enveloped_data_init(CMS_ContentInfo *cms)
+{
+    if (cms->d.other == NULL) {
+        cms->d.authEnvelopedData = M_ASN1_new_of(CMS_AuthEnvelopedData);
+        if (!cms->d.authEnvelopedData) {
```
nit : check against NULL:
`         if (cms->d.authEnvelopedData != NULL) { `

bukka replied:
It's actually `if (cms->d.authEnvelopedData == NULL) {`

Fixed now

FdaSilvaYY commented on crypto/cms/cms_env.c:
```diff
@@ -467,10 +556,9 @@ CMS_RecipientInfo *CMS_add0_recipient_key(CMS_ContentInfo *cms, int nid,
     CMS_KEKRecipientInfo *kekri;
-    env = cms_get0_enveloped(cms);
-    if (!env)
+    STACK_OF(CMS_RecipientInfo) *ris = CMS_get0_RecipientInfos(cms);
```
style: blank line after declaration.

bukka replied:
Fixed

FdaSilvaYY commented on doc/man3/CMS_encrypt.pod:
```diff
@@ -26,7 +27,8 @@ EVP_des_ede3_cbc() (triple DES) is the algorithm of choice for S/MIME use
 
 The algorithm passed in the B<cipher> parameter must support ASN1 encoding of
-its parameters.
+its parameters. If the cipher mode is GCM, then AuthEnvelopedData structure
```
nit:  `then **an** AuthEnvelopedData structure`

bukka replied:
Fixed

FdaSilvaYY commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+    ivl = EVP_CIPHER_CTX_iv_length(c);
+    if (ivl <= 0 || (iv = OPENSSL_malloc(ivl)) == NULL
+            || EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_GET_IV, ivl, iv) <= 0)
+        return -1;
```
possible leak of `iv` when `EVP_CIPHER_CTX_ctrl() return is <= 0`

bukka replied:
Fixed

bukka replied:
Good catch!

russhousley commented:
Thanks for all of this very valuable work.  The S/MIME 4.0 specifications are in the final stages, and they make use of AuthEnvelopedData!

Ideally, in the future, AAD support for the authenticated attributes in AuthEnvelopedData would be the next step.

bukka commented:
@russhousley Yes I plan to look on adding AAD support after this gets reviewed and possibly merged. I'm still thinking about the best API but the implementation should be fairly easy. Although I think it will require some improvements in the CMS testing as it's quite limited atm. I would like to look on adding ChaCha20-Poly1305 (thank you for creating RFC 8103 btw.) as well and I also thought about CCM support but it will probably require loading the whole plaintext to the memory BIO to get the length before it's processed by cipher BIO. That needs a bit more thinking!

bukka commented on crypto/cms/cms_env.c:
```diff
@@ -620,7 +709,10 @@ static int cms_RecipientInfo_kekri_encrypt(CMS_ContentInfo *cms,
 
-    ec = cms->d.envelopedData->encryptedContentInfo;
+    ec = cms_get0_env_enc_content(cms);
+    if (!ec) {
```
fixed

bukka commented on crypto/cms/cms_env.c:
```diff
@@ -121,6 +190,26 @@ CMS_ContentInfo *CMS_EnvelopedData_create(const EVP_CIPHER *cipher)
+{
+    CMS_ContentInfo *cms;
+    CMS_AuthEnvelopedData *aenv;
+    cms = CMS_ContentInfo_new();
```
Fixed

bukka commented on crypto/cms/cms_env.c:
```diff
@@ -467,10 +556,10 @@ CMS_RecipientInfo *CMS_add0_recipient_key(CMS_ContentInfo *cms, int nid,
-    if (!env)
+    STACK_OF(CMS_RecipientInfo) *ris = CMS_get0_RecipientInfos(cms);
+
+    if (!ris)
```
Fixed

bukka commented on crypto/cms/cms_env.c:
```diff
@@ -779,111 +874,208 @@ int CMS_RecipientInfo_encrypt(CMS_ContentInfo *cms, CMS_RecipientInfo *ri)
+
+    ok = 1;
+err:
+    if (tag != NULL)
```
Fixed

bukka commented on crypto/cms/cms_pwri.c:
```diff
@@ -42,16 +42,20 @@ CMS_RecipientInfo *CMS_add0_recipient_password(CMS_ContentInfo *cms,
+    STACK_OF(CMS_RecipientInfo) *ris;
     CMS_RecipientInfo *ri = NULL;
-    CMS_EnvelopedData *env;
+    CMS_EncryptedContentInfo *ec;
```
Not sure why me previous comment doesn't show here but it's used in cms_env.c quite a lot so I think it's better to have it consistent.

bukka replied:
As mentioned before, the `ec` is more consistent with the rest of the `cms_env.c` so resolving.

wolftobias commented:
@levitte Could you pls take a look at this? I also would need the support for 'AuthEnvelopData'.

wolftobias commented:
@bukka @russhousley @FdaSilvaYY @mattcaswell Hi guys, I think here is a 2nd reviewer needed! Who can support us?

wolftobias commented:
@bukka You developed this on the future openssl version 3.0.0, right?

mattcaswell commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+    i = ASN1_TYPE_get_octetstring_int(type, &tl, NULL, INT_MAX);
+    if (i <= 0 || (iv = OPENSSL_malloc(i)) == NULL)
+        return -1;
+    ASN1_TYPE_get_octetstring_int(type, &tl, iv, i);
```
Why do we do nothing with the taglen here?

bukka replied:
It's not really needed as it should be the same as `i`. That param is not really useful. I just added there to be consistent with `ASN1_TYPE_get_int_octetstring`.

mattcaswell commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+        return -1;
+
+    ivl = EVP_CIPHER_CTX_iv_length(c);
+    if (ivl <= 0 || (iv = OPENSSL_malloc(ivl)) == NULL)
```
Similarly this could be stack allocated.

bukka replied:
Fixed

mattcaswell commented on doc/man3/CMS_encrypt.pod:
```diff
@@ -26,7 +27,8 @@ EVP_des_ede3_cbc() (triple DES) is the algorithm of choice for S/MIME use
 The algorithm passed in the B<cipher> parameter must support ASN1 encoding of
-its parameters.
+its parameters. If the cipher mode is GCM, then an AuthEnvelopedData structure
+containing MAC is used. Otherwise an EnvelopedData structure is used.
```
Perhaps we should say other AEAD algorithms are not supported.

bukka replied:
Done. Not sure about the wording. Happy to change it if you prefer a different wording?

mattcaswell replied:
The wording looks fine.

mattcaswell commented on include/openssl/cms.h:
```diff
@@ -152,6 +152,7 @@ int CMS_decrypt_set1_password(CMS_ContentInfo *cms,
 STACK_OF(CMS_RecipientInfo) *CMS_get0_RecipientInfos(CMS_ContentInfo *cms);
 int CMS_RecipientInfo_type(CMS_RecipientInfo *ri);
 EVP_PKEY_CTX *CMS_RecipientInfo_get0_pkey_ctx(CMS_RecipientInfo *ri);
+CMS_ContentInfo *CMS_AuthEnvelopedData_create(const EVP_CIPHER *cipher);
```
This function needs documenting

bukka replied:
I agree. The only thing is that there is currently no documentation for `CMS_EnvelopedData_create` which should be probably also present. Do you think that it would make more sense if I created a new PR and added some docs for `CMS_EnvelopedData_create` and possibly some other CMS related functions? Then `CMS_AuthEnvelopedData_create` specific part would be added in this PR.

mattcaswell replied:
I'd be delighted if you opened a new PR and documented some of the other functions in another PR. Strictly all I need to approve this PR is the ensure that this new function is documented (whether or not the others are). I'll let you decide how best to tackle that.

bukka replied:
Added docs for `CMS_EnvelopedData_create` in #11529 . Will add the auth part into it once merged.

bukka replied:
Docs added.

mattcaswell commented on include/openssl/evp.h:
```diff
@@ -355,6 +355,10 @@ int (*EVP_CIPHER_meth_get_ctrl(const EVP_CIPHER *cipher))(EVP_CIPHER_CTX *,
+/* Set the IV used by the cipher */
+# define         EVP_CTRL_SET_IV                         0x27
+/* Get the tag length for AEAD cipher */
+# define         EVP_CTRL_AEAD_GET_TAGLEN                0x28
```
These ctrls need to be documented

bukka replied:
Will add it. I'm still not sure if we need `EVP_CTRL_SET_IV` as mentioned above. Will need to check it first.

mattcaswell replied:
So, I guess this is still outstanding.

bukka replied:
This has been removed as it's no longer needed.

mattcaswell commented on include/openssl/evp.h:
```diff
@@ -1141,6 +1145,8 @@ int EVP_CIPHER_asn1_to_param(EVP_CIPHER_CTX *c, ASN1_TYPE *type);
 int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type);
 int EVP_CIPHER_get_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type);
+int EVP_CIPHER_set_asn1_aead_params(EVP_CIPHER_CTX *c, ASN1_TYPE *type);
+int EVP_CIPHER_get_asn1_aead_params(EVP_CIPHER_CTX *c, ASN1_TYPE *type);
```
I think these should probably not be public functions.

bukka replied:
I thought that they are quite similar to `EVP_CIPHER_set_asn1_iv` resp. `EVP_CIPHER_get_asn1_iv` so I made that public but again happy to make them private.

mattcaswell replied:
Historically there were lots of functions that were made public that should not have been - so just because there are other similar public functions is probably not a good enough reason to make these ones public. I tend to default to making functions private unless I can see good reasons to make it a public API. Making things public makes it a lot more difficult to refactor code later if we need to. So, I think these should be made private.

bukka replied:
Ok. Done.

mattcaswell commented on crypto/asn1/evp_asn1.c:
```diff
@@ -95,21 +116,64 @@ int ASN1_TYPE_get_int_octetstring(const ASN1_TYPE *a, long *num,
+    return 0;
+}
+
+int ASN1_TYPE_get_octetstring_int(const ASN1_TYPE *a, long *num,
```
Its not clear to me why these two functions should sit in the generic ASN.1 sub-library. They don't seem especially useful except for a very specific purpose. Certainly I don't see them as public functions.

bukka replied:
They are almost the same like `ASN1_TYPE_get_int_octetstring` resp. `ASN1_TYPE_get_int_octetstring` except the order in ASN.1 sequence. The use is quite similar as well so I put them to the same file. Would you prefer a different file? The reason why they are public is for consistency with  `ASN1_TYPE_get_int_octetstring` resp. `ASN1_TYPE_get_int_octetstring` that are also public but happy to make them private if you think it makes more sense?

mattcaswell replied:
Ok, keep them where they are, but please make them private.

bukka replied:
Done

mattcaswell commented on test/recipes/80-test_cms.t:
```diff
@@ -239,6 +238,16 @@ my @smime_cms_tests = (
+      [ "-decrypt", "-in", "test.cms", "-out", "smtst.txt", "-inform", "PEM",
+	"-secretkey", "000102030405060708090A0B0C0D0E0F",
+	"-secretkeyid", "C0FEE0" ]
+    ],
```
I know your just following the way the other CMS tests are done, but really I'd like to see stronger tests that this. Rather than just testing things through the command line we should really be testing the API directly...e.g. with some negative tests too.

bukka replied:
I agree! I guess it might make more sense if I create a new with some tests for encryption and then just extend it in this PR. What do you think?

mattcaswell replied:
Again I'd be delighted if you want to do that. I would accept more minimal testing which just tests the newly added functions in order to approve this PR. I'll leave it with you to decide.

bukka replied:
I plan to add those tests and already started looking into it. It's basically next thing that I want to do before I start looking to the CCM and ChaCha20-Poly1305 in CMS as it's currently quite a pain to debug issues and API tests would really help.

I'm just thinking that it would be helpful for me if it doesn't have to be requirement for this PR? The reason is that I'm spending quite a bit of time just on rebasing it and integrating changes into it (e.g. the GOST related one took a bit of time). If it doesn't need to wait for adding a new way of testing in CMS, that would be great as it would save me some time!

bukka replied:
I just extended the `cmsapitest.c` to at least cover the gcm mode. It's not a great test but hope it should do for this PR.

I was thinking about improvements in CMS testing and I think that the main issue it that the actual ASN.1 elements are not verified in any way. It's basically just encrypted envelope can be decrypted and the result is the same. But if there was some structural bug in both, then it could be easily missed. Maybe some kind of match for asn1parse as part of cli test..? Anyway that's more for a different PR but thoughts on this would be appreciated.

bukka commented on crypto/cms/cms_env.c:
```diff
@@ -779,111 +874,208 @@ int CMS_RecipientInfo_encrypt(CMS_ContentInfo *cms, CMS_RecipientInfo *ri)
 
-    cms_env_set_originfo_version(env);
+    env->version = cms_env_get_oi_version(env->originatorInfo, env->version);
+    env->version = cms_env_get_ris_version(env->recipientInfos, env->version);
```
As mentioned above, there should be no issue so resolving.

bukka commented on crypto/cms/cms_enc.c:
```diff
@@ -74,12 +74,25 @@ BIO *cms_EncryptedContent_init_bio(CMS_EncryptedContentInfo *ec)
                 goto err;
-            piv = iv;
+            if (!(EVP_CIPHER_flags(ciph) & EVP_CIPH_FLAG_AEAD_CIPHER)
+                    || EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_SET_IV, ivlen,
```
I remember that it didn't work for GCM for some reason. It's been over 5 months so need to look a bit more. I see the code changed a bit since than as well so will take a proper look if it still doesn't work even with `EVP_CTRL_SET_IVLEN`.

bukka replied:
This has been changed and no ctrls are needed now.

bukka commented on crypto/cms/cms_env.c:
```diff
@@ -779,111 +875,207 @@ int CMS_RecipientInfo_encrypt(CMS_ContentInfo *cms, CMS_RecipientInfo *ri)
 
-    cms_env_set_originfo_version(env);
+    env->version = cms_env_get_oi_version(env->originatorInfo, env->version);
+    env->version = cms_env_get_ris_version(env->recipientInfos, env->version);
```
Yeah I'm sorry about this. I remember that I thought that there are different rules but after I refactored it and read again the RFC, I found out that it's actually the same so I just kept the refactored version as I thought that it's a bit more readable. Should I revert this refactoring?

mattcaswell replied:
Unless there is a good reason for changing this code, I'd prefer to keep the churn to a minimum. So, reverting this seems like the way to go.

bukka replied:
I have been looking into this again and when reading it again and figuring out the logic, the new version seems much clearer to me - I was able to get the logic of the new version generation quicker. So the main reason to keep it would be a better clarity of the code. Would you still prefer to put it back as it was? Potentially I could create a new PR for this later if that's better?

bukka replied:
After thinking a new PR with a proper test would be better so put it back as it was.

bukka commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+
+    if (type == NULL) {
+        return i;
+    }
```
Fixed

bukka commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+    }
+
+    i = ASN1_TYPE_get_octetstring_int(type, &tl, NULL, INT_MAX);
+    if (i <= 0 || (iv = OPENSSL_malloc(i)) == NULL)
```
Good point! Fixed!

bukka commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+
+    if (!(c->cipher->flags & EVP_CIPH_FLAG_AEAD_CIPHER)) {
+        return -1;
+    }
```
Fixed

bukka commented on crypto/evp/evp_lib.c:
```diff
@@ -114,6 +120,64 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+        return i;
+    }
+
+    if (!(c->cipher->flags & EVP_CIPH_FLAG_AEAD_CIPHER)) {
```
Fixed

bukka commented:
> @bukka You developed this on the future openssl version 3.0.0, right?

@wolftobias Yes it's on master. I just rebased it.

wolftobias commented:
I need this new feature on the current v1.1.1xxx, maybe I have to integrate it manually in my local build, version 1.1.1 has long time support till 2023 and version 3 has not target release date yet. Do you or anybody else see a chance to make this available for v1.1.1?

mattcaswell commented:
> Do you or anybody else see a chance to make this available for v1.1.1?

Unfortunately not. 1.1.1 is a stable release. We don't add new features to stable releases (and this is counted as a new feature).

wolftobias commented:
> > Do you or anybody else see a chance to make this available for v1.1.1?
> 
> Unfortunately not. 1.1.1 is a stable release. We don't add new features to stable releases (and this is counted as a new feature).

Yes I understand your point. Is there already an idea about the target release date of v3?

mattcaswell commented:
> Yes I understand your point. Is there already an idea about the target release date of v3?

We hope to be code complete by the end of this year, i.e. beta release around that point. With a possible release around early Q2 2020 (FIPS validation would come some time later).

wolftobias commented:
@bukka @mattcaswell What's the current status?

mattcaswell commented:
Oh...sorry, it looks like @bukka previously responded to earlier review comments and is waiting on me to respond to some questions. I'll try and answer those now.

mattcaswell commented:
I've responded to the various comments above...over to @bukka. Looks like this PR needs a rebase due to conflicts with master

bukka commented:
I have been bit busy lately. Will try to address all the outstanding bits when I have got a bit more free time.

wolftobias commented:
> I have been bit busy lately. Will try to address all the outstanding bits when I have got a bit more free time.

time is always against us! But it would be great if it would be possible for you to finish the open tasks, otherwise we're risking not to go in the next release and that will be sad. Also the reviewers took there time to work on this.

wolftobias commented:
@bukka I tried to test your unauthenticated attribute feature you developed, but the result is the error below. Could you please help me further?

C:\download\openssl-bukka\win64\bin>openssl cms -encrypt -in test.txt -out test.txt.enc -aes-128-gcm -secretkey 000102030405060708090A0B0C0D0E0F -secretkeyid C0FEE0
35672:error:060CD0E4:digital envelope routines:EVP_CIPHER_param_to_asn1:reason(228):crypto/evp/evp_lib.c:46:
35672:error:2E078066:CMS routines:cms_EncryptedContent_init_bio:cipher parameter initialisation error:crypto/cms/cms_enc.c:142:
35672:error:2E07F068:CMS routines:CMS_final:cms lib:crypto/cms/cms_smime.c:764:

bukka commented:
@wolftobias Sure I will get there eventually! :) Just pushed the part of making the functions private and plan to slowly address the rest. There is still lots of time before the release ;)

About your error, it looks like you might not be using the patched code but not really. I would suggest to wait until the patch is fully ready before testing it.

bukka commented on crypto/evp/evp_lib.c:
```diff
@@ -148,6 +155,50 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+    if (type == NULL)
+        return 0;
+
+    tl = EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_AEAD_GET_TAGLEN, 0, NULL);
```
@mattcaswell I have rebased the PR and started seeing issues likely caused by #9231 from @slontis as the GCM cipher is now moved to providers which currently returns `EVP_CTRL_RET_UNSUPPORTED` for this type. From the code, it looks that it's just a temporary thing (`default:` in the middle of switch and the providers params logic skipped) so I was wondering if I should just wait till this is addressed or I need to look for a different solution?

slontis replied:
I will look at this now - should be easy to fix..

slontis replied:
Are you supporting CCM also?

slontis replied:
See #9698

bukka replied:
Thanks for creating PR so quickly!

CCM is not currently supported by this PR as it's a bit more tricky when using BIO cipher and will require some further changes to get the input length which needs to be known in advance for CCM. Plan to look into it later.

slontis replied:
Yep the BIO's are not pretty :)

bukka replied:
This is no longer used so marking as resolved.

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -148,6 +155,50 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+        return -1;
+    asn1_type_get_octetstring_int(type, &tl, iv, i);
+
+    if (EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_AEAD_SET_IVLEN, i, NULL) <= 0
```
This is not the normal flow for setting the iv. This is normally done through a call to init and pass the other params as NULL. If it is possible I would prefer it worked that way. The flow is already complex enough without adding extra controls. Let me know if that works or not.

bukka replied:
I did it mainly because of the flow in `cms_EncryptedContent_init_bio` which is for decryption like this:

```
EVP_CipherInit_ex(ctx, ciph, NULL, NULL, NULL, enc)
EVP_CIPHER_asn1_to_param(...)
EVP_CipherInit_ex(ctx, NULL, NULL, ec->key, piv, enc) /* piv is NULL in this case */
```

The thing is that iv is retrieved from ASN1 inside `EVP_CIPHER_asn1_to_param` so I would need to remember it somewhere but that seemed a bit less clean than using ctrl. I might need to rethink though. If you have any suggestion what you think would be the cleanest way, happy to look into that!

slontis replied:
I was suggesting calling EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc) inside EVP_CIPHER_asn1_to_param(...). Or does that not work because you need a key first??

bukka replied:
Key usually comes from `CMS_EncryptedContentInfo`. Are you suggesting to use
`EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc)` inside `EVP_CIPHER_asn1_to_param` and then call it again in `cms_EncryptedContent_init_bio` with the key? Basically something like

```
EVP_CipherInit_ex(ctx, ciph, NULL, NULL, NULL, enc) // in cms_EncryptedContent_init_bio
EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc) // in EVP_CIPHER_asn1_to_param
EVP_CipherInit_ex(ctx, NULL, NULL, key, NULL, enc) //  in cms_EncryptedContent_init_bio
```

Also EVP_CIPHER_asn1_to_param is a public function so could it change the ctx behavior somehow? I mean if it will be always the case that `EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc)` will only set `iv` and doesn't do another initialization.

slontis replied:
Ahh yeah so that was a stupid idea :)
Ok another idea. Because this is new code that will not be in 1_1_1.
The new aes_gcm code will now generate an IV if you dont specify one (it was required for FIPS). So you shouldnt need to generate it yourself.. The only issue then would be if you need to get the iv back out in order to store it somewhere.. Try taking out the set_iv and see what happens.

newsuk-jzelenka replied:
I'm not sure if I follow. :) The described flow and this line of code is used just for decryption. In that case the iv comes from the ASN.1 params where it was stored when eveloped data has been created. How could it work if I take it out?

bukka replied:
I'm not sure if I follow. :) The described flow and this line of code is used just for decryption. In that case the iv comes from the ASN.1 params where it was stored when eveloped data has been created. How could it work if I take it out?

slontis replied:
Ok. I am talking about the set_iv ctrl that you added in the encrypt.. (obviously I need to go to sleep soon)

slontis replied:
The EVP_CTRL_AEAD_SET_IVLEN should be ok as it is an existing control (I am just looking at the 2 new ctrl id's you added in this PR). one is replaced by the PR above (to get the tag length) and the other one that sets the iv should be handled internally in openssl now.

slontis replied:
Sorry for the confusion.

bukka replied:
Ah yeah sorry I actually meant the code line below (EVP_CTRL_SET_IV) :). Anyway that's still used for decryption. Basically the decryption flow is `cms_EncryptedContent_init_bio -> if decryption -> EVP_CIPHER_asn1_to_param -> if GCM -> evp_cipher_get_asn1_aead_params`. This function parses the ASN.1, get the IV and set its length (this line) and then iv itself (line below). So basically I still need to set iv somehow :)

The encryption part is in `evp_cipher_set_asn1_aead_params` which sets tag length (EVP_CTRL_AEAD_GET_TAGLEN) and get iv (EVP_CTRL_GET_IV) so it can be stored in ASN.1.

slontis replied:
I will have to think about this a bit more. The interface is already confusing enough without yet another way of setting the iv into the ctx.

bukka replied:
Another option would be to implement it directly on the cipher using `set_asn1_parameters` resp. `get_asn1_parameters`. Then modifying of the context could be a bit more specific but could also result in duplicating the same logic when support for other AEAD ciphers is added in the future. I would probably have to think about it a bit more... Do you think that this would be potentially cleaner?

bukka replied:
I have done a different approach and wrap the functions so no need for ctrls.

slontis commented on crypto/evp/e_aes.c:
```diff
@@ -2584,14 +2584,22 @@ static int aes_gcm_ctrl(EVP_CIPHER_CTX *c, int type, int arg, void *ptr)
         memcpy(ptr, c->buf, arg);
         return 1;
 
+    case EVP_CTRL_AEAD_GET_TAGLEN:
```
EVP_CIPHER_CTX_tag_length() has been added - so this should not be necessary.

bukka replied:
This is no longer used so marking as resolved.

slontis commented on crypto/evp/e_aes.c:
```diff
@@ -2584,14 +2584,22 @@ static int aes_gcm_ctrl(EVP_CIPHER_CTX *c, int type, int arg, void *ptr)
         return 1;
 
+    case EVP_CTRL_SET_IV:
+        if (gctx->ivlen != arg)
```
If possible do this via an init() call instead.

bukka replied:
@slontis I need to remove those changes as I decided to wrap setting/getting ASN 1 params: https://github.com/openssl/openssl/pull/8024/files#diff-b5ce0bd68ed8ccf480748306416a7cccR85 which allows me to get rid of all the new ctrls. Still need to rebase it and fix some issue / address few other comments and should be ready for review.

bukka replied:
This is no longer used so marking as resolved.

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -180,10 +270,10 @@ CMS_RecipientInfo *CMS_add1_recipient_cert(CMS_ContentInfo *cms,
-    env = cms_get0_enveloped(cms);
-    if (!env)
+    ris = CMS_get0_RecipientInfos(cms);
+    if (!ris)
```
check with NULL

bukka replied:
Done

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -692,7 +785,10 @@ static int cms_RecipientInfo_kekri_decrypt(CMS_ContentInfo *cms,
 
-    ec = cms->d.envelopedData->encryptedContentInfo;
+    ec = cms_get0_env_enc_content(cms);
+    if (!ec) {
```
check with NULL

bukka replied:
Done

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -796,111 +892,207 @@ int CMS_RecipientInfo_encrypt(const CMS_ContentInfo *cms, CMS_RecipientInfo *ri)
+    ret = cms_EncryptedContent_init_bio(ec);
+
+    /* If error or no cipher end of processing */
+    if (!ret || !ec->cipher)
```
null check

bukka replied:
Done

slontis commented on crypto/cms/cms_pwri.c:
```diff
@@ -42,16 +42,20 @@ CMS_RecipientInfo *CMS_add0_recipient_password(CMS_ContentInfo *cms,
-    env = cms_get0_enveloped(cms);
-    if (!env)
+    ec = cms_get0_env_enc_content(cms);
+    if (!ec)
```
check with NULL

bukka replied:
Done

slontis commented on crypto/cms/cms_pwri.c:
```diff
@@ -42,16 +42,20 @@ CMS_RecipientInfo *CMS_add0_recipient_password(CMS_ContentInfo *cms,
+    if (!ec)
+        return NULL;
+    ris = CMS_get0_RecipientInfos(cms);
+    if (!ris)
```
check with NULL

bukka replied:
Done

bukka commented:
@slontis thanks for the review. I have still some failing cases that I need to debug to make test green and extend the tests as well as few doc addition. I have been a bit busy recently but will try to find some time to finelize this and address your comments. Will ping you once all ready!

bukka commented:
@slontis @mattcaswell I have addressed all comments and requested changes. Please could you review it again and let me know if there anything else to address.

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -1028,6 +1099,120 @@ BIO *cms_EnvelopedData_init_bio(CMS_ContentInfo *cms)
+
+    BIO_get_cipher_ctx(cmsbio, &ctx);
+
+    /* Do nothing when decrypting */
```
Can you expand this comment explaining why this is so.

slontis commented on crypto/err/openssl.txt:
```diff
@@ -249,6 +250,10 @@ CMS_F_CMS_ADD1_SIGNER:102:CMS_add1_signer
 CMS_F_CMS_ADD1_SIGNINGTIME:103:cms_add1_signingTime
 CMS_F_CMS_ADD1_SIGNING_CERT:181:cms_add1_signing_cert
 CMS_F_CMS_ADD1_SIGNING_CERT_V2:182:cms_add1_signing_cert_v2
+CMS_F_CMS_AUTHENVELOPEDDATA_CREATE:184:
```
The CMS_F values are no longer used in 3.0. So you can do CMSErr(0, r) now.
I guess this will make it slightly easier to backport to informal branches of 1.1.1

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -97,15 +148,16 @@ int EVP_CIPHER_param_to_asn1(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
  err:
     if (ret == -2)
-        EVPerr(EVP_F_EVP_CIPHER_PARAM_TO_ASN1, ASN1_R_UNSUPPORTED_CIPHER);
+        EVPerr(EVP_F_EVP_CIPHER_PARAM_TO_ASN1_EX, ASN1_R_UNSUPPORTED_CIPHER);
```
Just use 0 here,

slontis commented:
It looks like #11884 will need to be updated once this PR goes in.

slontis commented:
Can you verify the following are used as described:
https://tools.ietf.org/html/rfc5084 Section 3.2. GCMParameters

rfc5084:
authAttrs optionally contains the authenticated attributes.  The
CMS authenticated-data content type uses the same type to carry
authenticated attributes.  The authAttrs MUST be present if the
content type carried in EncryptedContentInfo is not id-data.

bukka commented:
@slontis Thanks for taking a look! Addressed your comments and hopefully removed all the `..._F_...` bits. ;)

> rfc5084:

the below is actually from rfc5083 :)

> authAttrs optionally contains the authenticated attributes. The
CMS authenticated-data content type uses the same type to carry
authenticated attributes. The authAttrs MUST be present if the
content type carried in EncryptedContentInfo is not id-data.

This is ok because `id-data` (`NID_pkcs7_data`) is always used for the authenticated-data content type. That matches the envelopedData that also support only `id-data`. This is set in https://github.com/openssl/openssl/pull/8024/files#diff-47b8520df0a2f8fa0433ff37611596f3R95-R96

bukka commented:
hmm that link to diff doesn't seem to work that well so this one should show it better: https://github.com/openssl/openssl/blob/9a45b84584440cd0aa37685c3d019c7c09d41a8c/crypto/cms/cms_env.c#L95-L96

bukka commented:
> I am not sure I follow the current ctrl mechanism related to EVP_CTRL_PROCESS_UNPROTECTED.
> I dont know how a user would add one of these attributes.
> If this is only for use by engines then that seems to be a weird interface to me.

I think the intention was probably for the engine use as it was added to support GOST ciphers. I assume it's for storing some cipher specific parameters.

I think that the user API should look differently. For encryption it could be similar like for setting recipient (e.g. after CMS_encrypt with CMS_STREAM) or a completely new function for the encryption. The decryption would probably require a new special CMS_decrypt variant as the authAttrs will need to be passed in.
 
> I also realize that the protected attributes are optional, but should still be implemented for this standard. (Even if it is in another PR).

I agree it would be useful. I would prefer another PR as it will be cleaner and might maybe cover setting unprotected attrs for enveloped data as well.

> So as far as I can see by manually inspecting the output cms data from a test it looks like it matches the standard. Ideally some interop tests against another toolkit would be advisable (maybe a java related toolkit?), otherwise we run the risk of outputting a format that may not be compatible when used with other toolkits. (Even if they were just manually done).

That's a good idea! Will play with AuthEnvelopedData in Bouncy Castle.

beldmit commented:
> I am not sure I follow the current ctrl mechanism related to EVP_CTRL_PROCESS_UNPROTECTED.
> I dont know how a user would add one of these attributes.
> If this is only for use by engines then that seems to be a weird interface to me.

It was introduced by me to process our (GOST) current CMS standard. Some of our ciphers (not really AEAD) support MAC calculation during the encryption/decryption process, and the resulting MAC goes to unprotected attributes.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

slontis commented:
@bukka did you get a chance to try some interop tests?

bukka commented:
@slontis I was moving house so was a bit busy but I started putting together a java cli app: https://github.com/bukka/jcrypto

It's a bit more generic so it can be extended and used for more things in the future. It took me a bit of time to figure out how to use BouncyCastle as I have never used it before and it's been actually some time since I wrote something in Java... :) I'm almost done with [CMS encryption with KEKRecipientInfo](https://github.com/bukka/jcrypto/blob/e1ee94ca9f85082d14df8ccf486c263ff9cdaa8a/src/main/java/eu/bukka/cms/CMSEnvelope.java#L27-L47) and hopefully will have something working so I can do some interop testing.

kaduk commented:
>  Also it seems to me that `AuthEnvelopedData` is not really that much used as I'm struggling to find any open source implementation so there is not much risk anyway.

For what it's worth, the IETF has been more active about ensuring that AuthEnvelopedData is used as new CMS-using documents come through the publication process.

t8m commented:
The travis failures are relevant.

bukka commented on util/libcrypto.num:
```diff
@@ -5286,3 +5286,5 @@ OSSL_PARAM_get_octet_string_ptr         ?	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_CTX_set_mac_key                ?	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_new_CMAC_key_with_libctx       ?	3_0_0	EXIST::FUNCTION:
+CMS_AuthEnvelopedData_create            ?	3_0_0	EXIST::FUNCTION:CMS
+CMS_AuthEnvelopedData_create_with_libctx ?	3_0_0	EXIST::FUNCTION:CMS
```
Fixed

bukka commented on crypto/evp/evp_lib.c:
```diff
@@ -104,15 +155,16 @@ int EVP_CIPHER_param_to_asn1(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
  err:
     if (ret == -2)
-        EVPerr(EVP_F_EVP_CIPHER_PARAM_TO_ASN1, ASN1_R_UNSUPPORTED_CIPHER);
+        EVPerr(0, ASN1_R_UNSUPPORTED_CIPHER);
```
Ah yeah thanks. Fixed.

bukka commented:
@t8m I forgot to do `make update`... Should be all fixed now. Thanks.

bukka commented:
@mattcaswell @slontis @t8m I think everything should be addressed. If there is any chance to get this to beta1, that would be awesome!

slontis commented on crypto/asn1/evp_asn1.c:
```diff
@@ -95,21 +117,64 @@ int ASN1_TYPE_get_int_octetstring(const ASN1_TYPE *a, long *num,
+    ASN1_OCTET_STRING *oct;
+    int32_t num;
+} asn1_oct_int;
+
```
Please add a reference to the documentation 
as well as a listing of the ans1 data that this is need for..

bukka replied:
Do you mean like a comment? This is basically just like `asn1_int_oct` above but swapped. I don't see any reference for `asn1_int_oct` elsewhere in the code.

slontis replied:
Yes a comment: I mean an RFC document and description of section where it is referenced from.
It makes it easier for someone looking at this later to know what it is doing.

bukka replied:
Pushed - please check if the format is fine?

bukka replied:
Ah it wasn't pushed but it's pushed now. :)

slontis commented on crypto/cms/cms_asn1.c:
```diff
@@ -245,6 +245,16 @@ ASN1_NDEF_SEQUENCE(CMS_EncryptedData) = {
         ASN1_IMP_SET_OF_OPT(CMS_EncryptedData, unprotectedAttrs, X509_ATTRIBUTE, 1)
 } ASN1_NDEF_SEQUENCE_END(CMS_EncryptedData)
 
+ASN1_NDEF_SEQUENCE(CMS_AuthEnvelopedData) = {
```
Add a reference to the RFC documentation and section that this is defined in..

bukka replied:
Do you mean just adding a comment. Something like `Defined in RFC 5084`?

newsuk-jzelenka replied:
it's actually RFC 5083 :)

slontis replied:
Yes.. But a little more details .. e.g RFC XXX "Section Blah : TitleX"

bukka replied:
Pushed - please let me know if the format is fine?

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -55,6 +83,25 @@ static CMS_EnvelopedData *cms_enveloped_data_init(CMS_ContentInfo *cms)
+            CMSerr(0, ERR_R_MALLOC_FAILURE);
+            return NULL;
+        }
+        cms->d.authEnvelopedData->version = 0;
```
Put a ref to documentation also to where this version  = 0 comes from

slontis replied:
I mean a reference to the section in the RFC here also..

slontis replied:
As an example of somewhere where I have done this for code..:
```
/*
 * Generate a SP800-56B RSA key.
 *
 * See SP800-56Br1 6.3.1 "RSA Key-Pair Generation with a Fixed Public Exponent"
 *    6.3.1.1 rsakpg1 - basic
 *    6.3.1.3 rsakpg1 - crt
```

bukka replied:
Ok this is defined as part for the type so added ref to RFC and section. Please let me know if the format is fine.

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -169,45 +235,34 @@ CMS_ContentInfo *CMS_EnvelopedData_create(const EVP_CIPHER *cipher)
-
-    cms_env_set_version(cms->d.envelopedData);
-    return 1;
+CMS_ContentInfo *CMS_AuthEnvelopedData_create(const EVP_CIPHER *cipher)
```
If this is a completely new API.. it probably doesn't need this variant at all..
The only reason the other ones are there is because they already existed.

bukka replied:
But this allows using `CMS_AuthEnvelopedData_create` without using the context which is much nicer for users IMO. This API is still useful and it's consistent with `CMS_EnvelopedData_create`.

slontis replied:
ok..

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -324,8 +379,8 @@ CMS_RecipientInfo *CMS_add1_recipient(CMS_ContentInfo *cms, X509 *recip,
-CMS_RecipientInfo *CMS_add1_recipient_cert(CMS_ContentInfo *cms,
-     X509 *recip, unsigned int flags)
+CMS_RecipientInfo *CMS_add1_recipient_cert(CMS_ContentInfo *cms, X509 *recip,
+                                          unsigned int flags)
```
indent is not right here.

bukka replied:
Fixed

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -774,7 +829,10 @@ static int cms_RecipientInfo_kekri_encrypt(const CMS_ContentInfo *cms,
 
-    ec = cms->d.envelopedData->encryptedContentInfo;
+    ec = cms_get0_env_enc_content(cms);
+    if (ec == NULL) {
```
NIT: remove {} for single lines.

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -843,7 +901,10 @@ static int cms_RecipientInfo_kekri_decrypt(CMS_ContentInfo *cms,
 
-    ec = cms->d.envelopedData->encryptedContentInfo;
+    ec = cms_get0_env_enc_content(cms);
+    if (ec == NULL) {
```
NIT: remove {} for single lines.

bukka replied:
Fixed

slontis commented on crypto/cms/cms_env.c:
```diff
@@ -1093,6 +1172,122 @@ BIO *cms_EnvelopedData_init_bio(CMS_ContentInfo *cms)
+     * The tag is set only for encryption. There is nothing to do for
+     * decryption.
+     */
+    if (!EVP_CIPHER_CTX_encrypting(ctx)) {
```
NIT: remove {} for single lines.

slontis commented on crypto/cms/cms_smime.c:
```diff
@@ -638,8 +638,11 @@ CMS_ContentInfo *CMS_encrypt_with_libctx(STACK_OF(X509) *certs,
+    cms = (EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER)
+          ? CMS_AuthEnvelopedData_create_with_libctx(cipher, libctx, propq)
+          : CMS_EnvelopedData_create_with_libctx(cipher, libctx, propq);
+    if (!cms)
```
cms == NULL

slontis commented on doc/man3/CMS_EnvelopedData_create.pod:
```diff
@@ -30,21 +44,23 @@ L<CMS_add0_recipient_key(3)>.
-but uses default values of NULL for the library context I<libctx> and the
-property query I<propq>.
+CMS_EnvelopedData_create() and  CMS_AuthEnvelopedData_create are similar to
+CMS_EnvelopedData_create_with_libctx() respective
```
and (instead of respective)

slontis commented on doc/man3/CMS_EnvelopedData_create.pod:
```diff
@@ -30,21 +44,23 @@ L<CMS_add0_recipient_key(3)>.
-structure it is usually not used in applications. The wrappers
-L<CMS_encrypt(3)> and L<CMS_decrypt(3)> are often used instead.
+Although CMS_EnvelopedData_create() and CMS_AuthEnvelopedData_create() allocate
+a new B<CMS_ContentInfo> structure, it is usually not used in applications.
```
it is => they are not usually used in...

slontis commented on test/cmsapitest.c:
```diff
@@ -99,7 +109,8 @@ int setup_tests(void)
 
-    ADD_TEST(test_encrypt_decrypt);
+    ADD_TEST(test_encrypt_decrypt_aes_cbc);
+    ADD_TEST(test_encrypt_decrypt_aes_gcm);
```
maybe add gcm 192 and 256 also..

bukka replied:
Added.

bukka commented:
@slontis Addressed your comments. Please let me know if the RFC references are fine.

slontis commented on crypto/asn1/evp_asn1.c:
```diff
@@ -95,21 +117,66 @@ int ASN1_TYPE_get_int_octetstring(const ASN1_TYPE *a, long *num,
+    int32_t num;
+} asn1_oct_int;
+
+/* This is used for AEAD algorithm params as the ones defined in
```
```
/* RFC 5084 - Section 2. Content-Authenticated Encryption Algorithms */
```
Would be enough

slontis replied:
If it is more than one line then it needs to be 
```
/*
 * XXXX
 * YYYYY
 */
```

newsuk-jzelenka replied:
Done

bukka replied:
Done

slontis commented on crypto/asn1/evp_asn1.c:
```diff
@@ -95,21 +117,66 @@ int ASN1_TYPE_get_int_octetstring(const ASN1_TYPE *a, long *num,
+    asn1_oct_int *atmp = NULL;
+    int ret = -1;
+
+    if ((a->type != V_ASN1_SEQUENCE) || (a->value.sequence == NULL)) {
```
another one to get rid of {}

newsuk-jzelenka replied:
Fixed

bukka replied:
Fixed

bukka commented:
@slontis Addressed the last 2 comments.

slontis commented:
Unless you really need to rebase could please start using fixups..
It makes it a lot harder to review your changes on something large if you do it via force pushing..

To do this you do
> git log
To view your commits and get the hash of the commit you want to add a fixup to..
>git commit --fixup hashid
>git push

Note you can also do this with the generated files (e.g libcrypto.num)..

slontis commented:
Have a look at something simple like
https://github.com/openssl/openssl/pull/12795/commits
And you be able to see the change fixups..
At the moment you are doing monolithic blobs - which means I have to scan everything each time (not just what you changed).

slontis commented:
NOTE: when we merge we the committers can easily squash these fixups as part of the process.

slontis reviewed (approved):
Approved assuming the CI loop passes.

I wish we had something to interop against though - your code is likely to become the thing that is interoped against in the future :)

bukka commented:
@slontis Thanks for the approval.

The reason why I kept it in a single commit is because I needed to rebase it quite a few times and there were lots of conflicts. If I had lots of commits, it would be really a pain to rebase it. Although I guess I could just merge master in if it's something that you are fine with? I can certainly do that next time. It works for me as soon as I don't need to rebase multiple commits...

slontis commented:
> The reason why I kept it in a single commit is because I needed to rebase it quite a few times and there were lots of 

I understand.. What I do is collapse it when it is ready to start getting reviews.. Then once people start reviewing just do the fixups to make it easy to review afte the initial review.. If it is going to sit there for a long time and cause you lots of conflicts later then collapse it again and start the fixups process again when it starts getting reviewed again.

t8m commented:
Approved, if the CI passes.

bukka commented:
@slontis @t8m @mattcaswell The CI is green so hopefully it's good to be merged.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

t8m commented:
In the force push I've rebased and resolved trivial conflict in libcrypto.num.
Merged to master.
Thank you for the contributed PR and for the patience with the reviews and merging.

elear commented:
@bukka does this require a PR for PHP?

bukka commented:
@elear No, it should just work.

lukaaash commented:
@bukka I think that a bug was introduced with this PR, in the following cms_asn1.c code:
```
/* Defined in RFC 5083 - Section 2.1. AuthEnvelopedData Type */
ASN1_NDEF_SEQUENCE(CMS_AuthEnvelopedData) = {
        ASN1_EMBED(CMS_AuthEnvelopedData, version, INT32),
        ASN1_IMP_OPT(CMS_AuthEnvelopedData, originatorInfo, CMS_OriginatorInfo, 0),
        ASN1_SET_OF(CMS_AuthEnvelopedData, recipientInfos, CMS_RecipientInfo),
        ASN1_SIMPLE(CMS_AuthEnvelopedData, authEncryptedContentInfo, CMS_EncryptedContentInfo),
        ASN1_IMP_SET_OF_OPT(CMS_AuthEnvelopedData, authAttrs, X509_ALGOR, 2),
        ASN1_SIMPLE(CMS_AuthEnvelopedData, mac, ASN1_OCTET_STRING),
        ASN1_IMP_SET_OF_OPT(CMS_AuthEnvelopedData, unauthAttrs, X509_ALGOR, 3)
} ASN1_NDEF_SEQUENCE_END(CMS_AuthEnvelopedData)
```

I believe that the two lines with authAttrs and unauthAttrs should actually look like this:
```
        ASN1_IMP_SET_OF_OPT(CMS_AuthEnvelopedData, authAttrs, X509_ATTRIBUTE, 1),
        ...
        ASN1_IMP_SET_OF_OPT(CMS_AuthEnvelopedData, unauthAttrs, X509_ATTRIBUTE, 2)
```

(It looks like the implicit IDs were copy&pasted from CMS_AuthenticatedData but not changed according to RFC 5083. Additionally, it seems that even the attributes in CMS_AuthenticatedData should use X509_ATTRIBUTE instead of X509_ALGOR.)

Ironically, it's quite fortunate that this bug exists - it causes OpenSSL to fail when it tries parsing AuthEnvelopedData that actually has authAttrs. Otherwise, it would likely fail on decryption because authAttrs is not used as AAD input for AES-GCM, as required by the RFC.

bukka commented:
@lukaaash Apology for the dealy (was busy at that time and then forgot). I just checked it and it is really a bug in the tag number as well as the type (I agree it should be X509_ATTRIBUTE for both). You are also right that this sort of failure is probably better than a bit cryptic failure in decryption that would likely happen due to logic in `ossl_cms_EncryptedContent_init_bio` which does not set AAD. 

There's currently no API for setting authAttrs and unauthAttrs so this might be a bit harder to test. It would be good to add some point too. I have been recently doing some work on BouncyCastle and the integration testing was mostly fine but haven't checked properly so will try to produce PEM there that we could use in the test. I think as a fix, it should just add veryfication for authAttr (setting it as AAD) when decrypting and fix those tags. In the future it would be good to also add support for attrs as well as making sure that authAttrs are set as AAD in encryption but that would be more a feature.

I will try to give a try in the coming weeks see if I can also produce a fix. I will also try to create a proper issue for this as this is closed.
